src/core/balance-sheet.ts

```diff
@@ -1,29 +1,55 @@
 import { getBalance as getBitcoindBalance } from "@services/bitcoind"
 import { lndsBalances } from "@services/lnd/utils"
 import { baseLogger } from "@services/logger"
-import { User } from "@services/mongoose/schema"
+import { InvoiceUser, User } from "@services/mongoose/schema"
 import { ledger } from "@services/mongodb"
 
 import { WalletFactory } from "./wallet-factory"
 
 const logger = baseLogger.child({ module: "balanceSheet" })
 
+export const updatePendingLightningTransactions = async () => {
+  let user, userWallet
+
+  // select distinct user ids from pending invoices
+  const usersWithPendingInvoices = InvoiceUser.aggregate([
+    { $match: { paid: false } },
+    { $group: { _id: "$uid" } },
+  ])
+    .cursor({ batchSize: 100 })
+    .exec()
+
+  for await (const { _id } of usersWithPendingInvoices) {
+    logger.trace("updating pending invoices for user %o", _id)
+    user = await User.findOne({ _id })
+    userWallet = await WalletFactory({ user, logger })
+    await userWallet.updatePendingInvoices()
+  }
+
+  const accounts = ledger.getAccountsWithPendingTransactions({ type: "payment" })
+  for await (const account of accounts) {
+    logger.trace("updating pending payments for account %o", account)
+    user = await User.findOne({ _id: ledger.resolveAccountId(account) })
+    userWallet = await WalletFactory({ user, logger })
+    await userWallet.updatePendingPayments()
+  }
+}
+
 export const updateUsersPendingPayment = async ({
   onchainOnly,
 }: { onchainOnly?: boolean } = {}) => {
-  let userWallet
-
-  for await (const user of User.find({})) {
-    logger.trace("updating user %o", user._id)
+  if (!onchainOnly) {
+    await updatePendingLightningTransactions()
+  }
 
-    // A better approach would be to just loop over pending: true invoice/payment
+  let userWallet
+  const users = User.find({ onchain: { $exists: true, $not: { $size: 0 } } }).cursor({
+    batchSize: 100,
+  })
```

Comment:
@krtk6160 could you please check this query in production? (just to validate if it return less than total Users)

Reply:
It returns only slightly (1.3%) less than total user count

Reply:
:) better than nothing

Reply:
@jrbuhl93 there are some optimization we could do here. only generate a new onchain address is a new user go to the receive bitcoin and swipe right to see the onchain address.

Reply:
it would result in a slightly worse tradeoff because the onchain address would need to be fetch at this time, but we would drastically reduce the number of onchain addresses we have to look up for in the long run